src/frontend/Semantic_error.ml

```diff
@@ -597,327 +601,302 @@ module StatementError = struct
-    | FuncDeclRedefined (name, ut, stan_math) ->
-        Fmt.pf ppf "Function '%s' %s signature %a" name
-          (if stan_math then "is already declared in the Stan Math library with"
-           else "has already been declared for")
-          UnsizedType.pp ut
-    | FunDeclExists name ->
-        Fmt.pf ppf
-          "Function '%s' has already been declared. A definition is expected."
-          name
-    | FunDeclNoDefn name ->
-        Fmt.pf ppf "Function '%s' is declared without specifying a definition."
-          name
-    | FunDeclNeedsBlock ->
-        Fmt.pf ppf "Function definitions must be wrapped in curly braces."
-    | NonRealProbFunDef ->
-        Fmt.pf ppf
-          "Real return type required for probability functions ending in \
-           _lpdf, _lupdf, _lpmf, _lupmf, _cdf, _lcdf, or _lccdf."
-    | ProbDensityNonRealVariate (Some ut) ->
-        Fmt.pf ppf
-          "Probability density functions require real variates (first \
-           argument). Instead found type %a."
-          UnsizedType.pp ut
-    | ProbDensityNonRealVariate _ ->
-        Fmt.pf ppf
-          "Probability density functions require real variates (first \
-           argument)."
-    | ProbMassNonIntVariate (Some ut) ->
-        Fmt.pf ppf
-          "Probability mass functions require integer variates (first \
-           argument). Instead found type %a."
-          UnsizedType.pp ut
-    | ProbMassNonIntVariate _ ->
+    | IllTypedAssignment (Operator.Equals, lt, rt) ->
         Fmt.pf ppf
-          "Probability mass functions require integer variates (first \
-           argument)."
-    | DuplicateArgNames ->
-        Fmt.pf ppf "All function arguments must have distinct identifiers."
-    | IncompatibleReturnType ->
+          "Ill-typed arguments supplied to assignment operator =:@ @[<v2>The \
+           left hand side has type@ @[%a@]@]@ @[<v2>and the right hand side \
+           has type@ @[%a@]@]"
+          UnsizedType.pp lt UnsizedType.pp rt
+    | IllTypedAssignment (op, lt, rt) ->
         Fmt.pf ppf
-          "Function bodies must contain a return statement of correct type in \
-           every branch."
+          "@[<v>Ill-typed arguments supplied to assignment operator %a=:@ \
+           @[<v2>The left hand side has type@ @[%a@]@]@ @[<v2>and the right \
+           hand side has type@ @[%a@]@]@ Available signatures for given \
+           lhs:@]@ %a"
+          Operator.pp op UnsizedType.pp lt UnsizedType.pp rt
+          SignatureMismatch.pp_math_lib_assignmentoperator_sigs (lt, op)
 end
 
-type t =
-  | TypeError of Location_span.t * TypeError.t
-  | IdentifierError of Location_span.t * IdentifierError.t
-  | ExpressionError of Location_span.t * ExpressionError.t
-  | StatementError of Location_span.t * StatementError.t
+type err =
+  | TypeError of TypeError.t
+  | IdentifierError of IdentifierError.t
+  | ExpressionError of ExpressionError.t
+  | StatementError of StatementError.t
+
+type t = Location_span.t * err
```

Comment:
It's a bit unsightly that `Semantic_error.t` is now an opaque tuple (opaque since users are expected to call our `location` instead `fst` directly) and `pp` only prints half of it.

Reply:
pp only printed half of of the actual content of the type before, it was just one level deeper.

I played around with having pp do the entire job, but it isn't clear where the shared code between Semantic_error and Syntax_error for printing the context would go. It can't stay in Errors, since that would introduce a circular dependency, and it can't be moved entirely into Middle.Location, since it relies on Frontend.Include_files